service/common/src/main/java/org/apache/polaris/service/admin/PolarisAdminService.java

```diff
@@ -468,22 +467,30 @@ private void authorizeGrantOnTableLikeOperationOrThrow(
         catalogRoleName);
     ResolverStatus status = resolutionManifest.resolveAll();
 
+    final PolarisEntitySubType searchSubtype;
+    if (subTypes.size() > 1) {
+      searchSubtype = PolarisEntitySubType.ANY_SUBTYPE;
+    } else {
+      searchSubtype = subTypes.getFirst();
+    }
+
     if (status.getStatus() == ResolverStatus.StatusEnum.ENTITY_COULD_NOT_BE_RESOLVED) {
       throw new NotFoundException("Catalog not found: %s", catalogName);
     } else if (status.getStatus() == ResolverStatus.StatusEnum.PATH_COULD_NOT_BE_FULLY_RESOLVED) {
       if (status.getFailedToResolvePath().getLastEntityType() == PolarisEntityType.TABLE_LIKE) {
-        if (subType == PolarisEntitySubType.ICEBERG_TABLE) {
-          throw new NoSuchTableException("Table does not exist: %s", identifier);
-        } else {
-          throw new NoSuchViewException("View does not exist: %s", identifier);
-        }
+        CatalogHandler.throwNotFoundExceptionForTableLikeEntity(identifier, searchSubtype);
       } else {
         throw new NotFoundException("CatalogRole not found: %s.%s", catalogName, catalogRoleName);
       }
     }
 
     PolarisResolvedPathWrapper tableLikeWrapper =
-        resolutionManifest.getResolvedPath(identifier, PolarisEntityType.TABLE_LIKE, subType, true);
+        resolutionManifest.getResolvedPath(
+            identifier, PolarisEntityType.TABLE_LIKE, searchSubtype, true);
```

Comment:
I'd recommend to add a override of `resolutionManifest.getResolvedPath()` to take a list of subtypes.